### DISCUSSION TITLE
Nvim-tree lazy loading error - Not an editor command: NvimTreeToggle

### DISCUSSION THREAD

**Discussion by EmreDogann** [nvim-tree/nvim-tree.lua#2298] [Q&A]
Hey there,

I have a problem when trying to lazy load nvim-tree while also loading a session at startup.
Below is my nvim-tree config:
```lua
local function my_on_attach(bufnr)
	local api = require("nvim-tree.api")

	local function opts(desc)
		return { desc = "nvim-tree: " .. desc, buffer = bufnr, noremap = true, silent = true, nowait = true }
	end

	-- default mappings
	api.config.mappings.default_on_attach(bufnr)

	-- custom mappings
	vim.keymap.set("n", "<C-t>", api.tree.change_root_to_parent, opts("Up"))
	vim.keymap.set("n", "?", api.tree.toggle_help, opts("Help"))
	vim.keymap.set("n", "h", api.node.navigate.parent_close, opts("Close Directory"))
	vim.keymap.set("n", "l", api.node.open.edit, opts("Open"))

	vim.keymap.set("n", "]e", "", { buffer = bufnr })
	vim.keymap.del("n", "[e", { buffer = bufnr })
	vim.keymap.set("n", "]d", api.node.navigate.diagnostics.next, opts("Next Diagnostic"))
	vim.keymap.set("n", "[d", api.node.navigate.diagnostics.prev, opts("Prev Diagnostic"))
end

-- Using lazy.nvim
return {
	"nvim-tree/nvim-tree.lua",
	lazy = true,
	dependencies = {
		"nvim-tree/nvim-web-devicons",
		"antosha417/nvim-lsp-file-operations",
	},
	cmd = {
		"NvimTreeOpen",
		"NvimTreeToggle",
		"NvimTreeFocus",
	},
	keys = {
		{ "<leader>e", ":NvimTreeToggle<CR>", silent = true },
	},
	config = function()
		require("nvim-tree").setup({
			on_attach = my_on_attach,
			disable_netrw = true,
			hijack_cursor = true,
			sync_root_with_cwd = true,
			diagnostics = {
				enable = true,
				show_on_dirs = true,
			},
			renderer = {
				highlight_opened_files = "all",
				highlight_git = true,
				indent_markers = {
					enable = true,
				},
			},
			update_focused_file = {
				enable = true,
				update_root = true,
				ignore_list = {},
			},
			view = {
				width = 35,
				side = "right",
				hide_root_folder = false,
				number = false,
				relativenumber = false,
			},
			modified = {
				enable = true,
				show_on_dirs = true,
				show_on_open_dirs = true,
			},
			git = {
				enable = true,
				ignore = false,
				timeout = 500,
			},
			filters = {
				dotfiles = false,
			},
		})

		-- Closes nvim-tree buffer when exiting neovim
		vim.api.nvim_create_autocmd({ "BufUnload" }, {
			pattern = "NvimTree*",
			callback = function()
				require("nvim-tree.api").tree.close()
			end,
		})
	end,
}
```

When I press `<leader>e` after loading an existing session, I get the following message:
```lua
E492: Not an editor command: NvimTreeToggle
```

In the following scenario:
1. Open a new nvim instance.
2. Create a session via :mksession (I use nvim-possession as my session manager which does under the hood).
3. Press `<leader>e` to toggle nvim-tree. It should then lazy load and appear.
4. Save session and exit nvim.
5. Relaunch nvim and load session.
6. Press `<leader>e` again, this time the error above occurs.

After investigating a bit, I found the culprit to be some global nvim-tree variables being saved into the session file.
I have `set sessionoptions+=globals` as some plugins require it to function. However, this also leads to the following global variables being saved to the session file:
```lua
...
let NvimTreeSetup =  1 
let NvimTreeRequired =  1 
...
```

For some reason, these are causing conflicts in the lazy loading of nvim-tree. If I manually set `NvimTreeSetup = 0`, then upon reloading the session, `<leader>e` works fine, until the next time the session is loading, at which point NvimTreeSetup is set back to 1 and I get the same errors again.

What can be done about this? Should I just manually set the global variable to 0 upon loading the session via a session load hook? Kinda hacky but could work I guess.

**Comment by alex-courtis** [comment]
Startup is difficult, due to the somewhat nondeterministic ordering of startup behaviour. The lazy plugin and session loading makes things greatly more difficult.

Perhaps this might be of assistance: [wiki: Workaround when using rmagatti/auto-session](https://github.com/nvim-tree/nvim-tree.lua/wiki/Recipes#workaround-when-using-rmagattiauto-session)

It may be pragmatic to avoid lazy loading for nvim-tree. There is no performance or resource benefit to lazy loading; nvim-tree does nothing but parse/set you configuration on `.setup({})`. Work is done only when the tree is opened.

**Reply by EmreDogann** [reply]
Ah I see, in that case, I shall do that. What I've had up to this point is setting `vim.g.NvimTreeSetup = 0` after the session load which seems to work perfectly, but I will try your suggestion. Thanks!

**Reply by alex-courtis** [reply]
I've added some notes: [wiki: Installation: Lazy Loading](https://github.com/nvim-tree/nvim-tree.lua/wiki/Installation#lazy-loading)

**Comment by xian-wen** [comment]
nvim-tree takes 500 ms+, which makes the startup of neovim a little bit slow, so I prefer to lazily load it. I think this works:
```lua
return {
  "nvim-tree/nvim-tree.lua",
  lazy = true,
  opts = {},
  keys = {
    {
      "<Leader>e",
      function ()
        require("nvim-tree.api").tree.toggle({ find_file = true })
      end,
      desc = "Explorer NvimTree (Dir of Current File)",
    },
    {
      "<Leader>E",
      function ()
        require("nvim-tree.api").tree.toggle({ path = vim.uv.cwd() })
      end,
      desc = "Explorer NvimTree (cwd)",
    },
  },
}
```
And running `:mks! session.nvim` before quit, then loading the session with `nvim -S session.nvim`, and pressing `<Leader>e`, no error occurs. By the way, I also have `globals` in `sessionoptions`.

**Reply by alex-courtis** [reply]
[Lazy loading](https://github.com/nvim-tree/nvim-tree.lua/wiki/Installation#lazy-loading) is not recommended.

nvim-tree startup (i.e. call to setup) does almost nothing and should return immediately.

Please [enable profiling](https://github.com/nvim-tree/nvim-tree.lua/wiki/Troubleshooting#performance-issues) and provide the logs so that we may resolve the 500ms startup problem.
